### ISSUE TITLE
Command `package:discovery` requires database connection

### ISSUE THREAD

**Issue by matteo-palazzo** [OFFLINE-GmbH/oc-mall-plugin#1007]
Ciao! As mentioned in the title, the `package:discovery` command requires a database connection due to `Notification::getEnabled()` method inside the  [`OFFLINE\Mall\Classes\Events\MailingEventHandler:26`](https://github.com/OFFLINE-GmbH/oc-mall-plugin/blob/cbe76f97bf1d796f93b2b97819c5aff132534a6b/classes/events/MailingEventHandler.php#L26) constructor. Let's say we clone an existing OctoberCMS + Mall repository and try to install the dependencies via composer. This will end up in a `SQLSTATE[HY000] [1045] Access denied` error, because the database is not configured yet.

Not a big deal, but a bit annoying. Should we address this issue?

@Mattia-Marchiorato

**Comment by tobias-kuendig** [comment]
This does sound annoying! 

I see two solutions:

- wrapping the constructor in a try/catch
- adding a `init` method that fetches the notifications and is called when binding the event handler into the app container here:
https://github.com/OFFLINE-GmbH/oc-mall-plugin/blob/cbe76f97bf1d796f93b2b97819c5aff132534a6b/classes/registration/BootEvents.php#L42

Something like:

```php
        $this->app->bind(MailingEventHandler::class, fn() => (new MailingEventHandler)->init());
        $this->app['events']->subscribe(MailingEventHandler::class);
```

I would prefer the second option, what's your opinion?

**Comment by matteo-palazzo** [comment]
Thanks Tobias for the reply!

I think the second option will not solve the issue. We should prevent a database call during the binding, otherwise it will throw the same error.

Maybe we should add a try/catch in the `Notification::getEnabled()` method. Something like:

```php
public static function getEnabled()
{
    try {
        $notifications = Notification::query()->where('enabled', true)->get()->pluck('template', 'code');
    } catch (\Exception $exception) {
        $notifications = collect([]);
    }

    return Cache::rememberForever(self::CACHE_KEY, fn() => $notifications);
}
```

Not a big fan. Catching this kind of exception sounds a bit "wrong" to me. But it's just my opinion! Still in the same function [we check the `offline_mall_notifications` table existence](https://github.com/OFFLINE-GmbH/oc-mall-plugin/blob/cbe76f97bf1d796f93b2b97819c5aff132534a6b/models/Notification.php#L34), and return empty collection otherwise. Maybe we should check if the application is running in the console and, if so, return an empty collection. Something like:

```php
public static function getEnabled()
{
    if (App::runningInConsole()) return collect([]);

    return Cache::rememberForever(self::CACHE_KEY, function () {
        return Notification::where('enabled', true)->get()->pluck('template', 'code');
    });
}

```
Does this mean we'll not send notifications if Laravel is bootstrapped in artisan command? Not sure if it's a great solution. The third option would be to rewrite the `MailingEventHandler` in order to retrieve the enabled notifications where required instead of in the constructor. Something like:

```php
public function checkoutSucceeded($result)
{
    $enabledNotifications = Notification::getEnabled();
    
    if ($enabledNotifications->has('offline.mall::checkout.succeeded')) {
        // Magic things...
    }
}
```
A bit redundant, but I don't see other cons. What's your opinion? Recap:

- Try/catch in `getEnabled()` method.
- Check `App::runningInConsole()`.
- Refactor `MailingEventHandler` class.

**Comment by tobias-kuendig** [comment]
Thank you for the details @matteo-palazzo, I went with the easiest method that should fix the issue, available in v3.0.17 (c161543)
